### ISSUE TITLE
Spring Boot app crashes when using splunkt-otel-java and spring-kafka

### ISSUE THREAD

**Issue by konradrenner** [signalfx/splunk-otel-java#1456]
I am trying to get the splunk-otel-javaagent working with a Spring Boot application. Everything works fine, as long as I am not using spring-kafka. If I am using the splunk otel javaagent alongside with spring-kafka I get the following error:

`Caused by: org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'io.micrometer.core.instrument.MeterRegistry' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {}`

**Important Sidenote**: It is wirking when I am using the upstream opentelemetry agent.

I am using the following code to intialize OTel Registry (copied from OTel Spring example):

  ```
@Bean
    @ConditionalOnClass(name = "io.opentelemetry.javaagent.OpenTelemetryAgent")
    public MeterRegistry otelRegistry() {
        Optional<MeterRegistry> otelRegistry = Metrics.globalRegistry.getRegistries().stream()
                .filter(r -> r.getClass().getName().contains("OpenTelemetryMeterRegistry"))
                .findAny();
        otelRegistry.ifPresent(Metrics.globalRegistry::remove);
        return otelRegistry.orElse(null);
    }
```

I am using the following dependencies:

```
<groupId>org.springframework.boot</groupId>
<artifactId>spring-boot-starter-parent</artifactId>
<version>3.0.6</version>
```

```
<dependency>
 <groupId>io.micrometer</groupId>
<artifactId>micrometer-tracing-bridge-otel</artifactId>
</dependency>
```

```
<dependency>
<groupId>org.apache.kafka</groupId>
<artifactId>kafka-streams</artifactId>
</dependency>
<dependency>
 <groupId>org.springframework.kafka</groupId>
<artifactId>spring-kafka</artifactId>
</dependency>
```

```
<groupId>com.splunk</groupId>
<artifactId>splunk-otel-javaagent</artifactId>
<version>1.28.0</version>
```

**Comment by laurit** [comment]
Currently, by default, splunk agent disables upstream micrometer instrumentation so there probably isn't a meter registry with class `OpenTelemetryMeterRegistry`.

**Comment by konradrenner** [comment]
Hi, thank you for the information. Is there a way to enable it?

**Comment by laurit** [comment]
It is possible to enable it. This instrumentation bridges micrometer metrics to otel metrics. Currently, by default, splunk agent does not emit otel metrics, but rather uses a bundled micrometer to send the metrics. That is why splunk agent disables the upstream instrumentation and has its own instrumentation that sends the metrics from application micrometer instance to the one bundled in the agent.

**Comment by konradrenner** [comment]
Hi, thanks for your answer, but I am confused now. What do I have to do, to get things working? 
For metrics I already use this one:
<dependency>
    <groupId>io.micrometer</groupId>
    <artifactId>micrometer-registry-signalfx</artifactId>
</dependency>

**Comment by laurit** [comment]
You should consider whether you need the `otelRegistry` method at all when running with splunk agent, perhaps adding `@ ConditionalOnMissingClass("com.splunk.opentelemetry.javaagent.SplunkAgent")` would work?
What do you expect to happen, metrics being sent from `micrometer-registry-signalfx` that is inside the application? Metrics being sent from the micrometer that is bundled with the agent? Metrics being sent over otlp?

**Comment by konradrenner** [comment]
Unfortunately, that doesn't work either. I think the problem is that when using the OTel upstream Java agent, the following class is available at runtime: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/micrometer/micrometer-1.5/library/src/main/java/io/opentelemetry/instrumentation/micrometer/v1_5/OpenTelemetryMeterRegistry.java , which is not available in the case of the Splunk OTel agent.

I only added the micrometer-signalfx dependency as a test (I had the micrometer-otlp in before, which gives the same error). Maybe to understand why I need this in the first place: I don't only want to send technical metrics, but also business metrics generated in the code. Actually, I would prefer to use the OpenTelemetry APIs directly, but according to the Spring Boot documentation, Micrometer should be used because it is deeply integrated.

Basically, the upstream OTel agent works for the project. However, the problem is that I do not get the profiling functions of Splunk O11y Cloud with the upstream OTel agent.

**Comment by konradrenner** [comment]
One more small addition on the subject of micrometer: Quarkus also relies on the Micrometer APIs for metrics and not directly on the OpenTelemetry APIs, so this is not an exclusive Spring Boot issue.

With Quarkus, there is the additional small problem that there is no Splunk OTel agent plugin (just a plugin for OTel upstream agent) and therefore the profiling feature for this runtime is not available in Splunk O11y Cloud. But that is another topic :)

**Comment by laurit** [comment]
> Unfortunately, that doesn't work either. I think the problem is that when using the OTel upstream Java agent, the following class is available at runtime: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/micrometer/micrometer-1.5/library/src/main/java/io/opentelemetry/instrumentation/micrometer/v1_5/OpenTelemetryMeterRegistry.java , which is not available in the case of the Splunk OTel agent.

I had a closer look at the code and I think the issue is that splunk agent adds a metics registry similarly to `OpenTelemetryMeterRegistry` it just happens to have a different class name, try changing the filter to also include classes that contain `ApplicationMeterRegistry`.

**Comment by konradrenner** [comment]
Do I also have to change the following one:
@ConditionalOnClass(name = "io.opentelemetry.javaagent.OpenTelemetryAgent")
?

**Comment by konradrenner** [comment]
Tried to change it to:
`@Bean
    @ConditionalOnClass(name = "com.splunk.opentelemetry.javaagent.SplunkAgent")
    public MeterRegistry otelRegistry() {
        Optional<MeterRegistry> otelRegistry = Metrics.globalRegistry.getRegistries().stream()
                .filter(r -> r.getClass().getName().contains("ApplicationMeterRegistry"))
                .findAny();
        otelRegistry.ifPresent(Metrics.globalRegistry::remove);
        return otelRegistry.orElse(null);
    }`
but doesn't work either...

**Comment by konradrenner** [comment]
Great! Thank you for your help, now it is working. If you like, I can contribute the example. If you want, you just have to tell me how or where you document something like that :)

**Comment by laurit** [comment]
We plan to move from micrometer based metrics to otel metrics following the release of `opentelemetry-java-instrumentation` v2 which will hopefully happen in the next few months. After that splunk agent will behave the same way as otel agent in this regard.

**Comment by konradrenner** [comment]
Great news, looking forward to it :+1:
